### DISCUSSION TITLE
tutorials/best_practices/version_control_systems

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#162] [User-contributed notes]
# tutorials/best_practices/version_control_systems

Introduction: Godot aims to be VCS-friendly and generate mostly readable and mergeable files. Version control plugins: Godot also supports the use of version control systems in the editor itself. H...

https://docs.godotengine.org/en/stable/tutorials/best_practices/version_control_systems.html

**Comment by buanzo** [comment]
It is of particular importance to remind developers of large file tracking. Some might find tools like git annex useful, whilst others might prefer to use large-file support (LFS) in commercial providers. I would ask Godot team to advice on the matter.

**Comment by RyanMan56** [comment]
I agree it would be useful to have some guidance from the Godot team of large file tracking approaches. In case anyone is looking for suggestions in the meantime this is my approach using git-lfs on mac. My repo is hosted on GitHub but it should be the same process for other providers:

1. `brew install git-lfs`
2. In your repo's root folder run `git lfs install`
3. 
```
git lfs track "*.png"
git lfs track "*.jpg"
git lfs track "*.wav"
git lfs track "*.ogg"
git lfs track "*.glb"
git lfs track "*.gltf"
git lfs track "*.obj"
git lfs track "*.fbx"
git lfs track "*.dae"
git lfs track "*.blend"
git lfs track "*.tscn"
```
(Add any other extensions of large files you are using. This is just my list)

4. `git add .`
5. `git commit -m "configured lfs"`
6. `git push`

**Reply by RyanMan56** [reply]
If you're using windows or linux check the git-lfs installation instructions here: https://git-lfs.com/

**Reply by Calinou** [reply]
I would also track `*.exr` and `*.lmbake` for baked LightmapGI data (plus `*.hdr` for HDRI panorama skies). These can grow large in complex scenes, and you'll be frequently overwriting them.

I wouldn't track `*.tscn` though as it's a text-based format, only `*.scn` and `*.res` (binary scene/resource formats).

A few other extensions that you could add: `*.mp3`, *`.dds`, `*.ktx`

**Reply by Anuril** [reply]
It's important to **not track `*.tscn`** as you won't be able to resolve merge conflicts. 

This is what's the content of a .tscn file if there is a merge conflict which git tries to resolve: 
```
version https://git-lfs.github.com/spec/v1
<<<<<<< HEAD
oid sha256:c1ac8dcfe11e24a1a5c5dec27170db7125e007819b1957caa03ec77deb0c42a6
size 17082
=======
oid sha256:e73773647a7d490984dbf6e80893f0eeab2697a9595427e75c60b6f49e6bc7ce
size 21420
>>>>>>> bfc7007 (Added new Road Tileset)
```

**Comment by buanzo** [comment]
Good. I don't think GitHub's LFS is a good fit for security reasons, but it
makes sense to provide that pointer.

I am currently solving my issues with a pre commit hook.

**Reply by RyanMan56** [reply]
What are the security reasons? Is it because you can't verify what someone else on your team has committed to LFS?

**Comment by buanzo** [comment]
GitHub LFS only works with username+password authentication, and not with
pubkey

**Reply by RyanMan56** [reply]
I'm using GitHub LFS with an ssh key, not username and password

**Comment by buanzo** [comment]
I'd really like to know how, as the GitHub LFS ssh helper does not allow
it. it might be a good documentation point, although I assume Godot would
prefer to keep the documentation agnostic

**Reply by RyanMan56** [reply]
I didn't do anything special, I just set up my ssh key as usual following the official GitHub docs:
- [Generating a new SSH key and adding it to the ssh-agent](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent?platform=mac)
- [Adding a new SSH key to your GitHub account](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account?platform=mac)

And then set up git-lfs using these steps from my first comment:
1. `brew install git-lfs`
2. `git lfs install` in my project's root folder

**Comment by buanzo** [comment]
Well, that's exactly what I've done. And I use GitHub with a myriad of ssh
keys for dozens of repos. And only the ones with LFS end up failing.
Setting git trace and taking the output for a good ole google spin raises
the same response: GitHub LFS's helper needs to use http auth at an
instance, so it's not pure ssh pubkey based

**Reply by Calinou** [reply]
Please use replies to the above comments instead of creating new comments, as people don't know what you're replying to otherwise. There's a comment box below each comment for this reason :slightly_smiling_face:

**Reply by RyanMan56** [reply]
I've just done a bit and digging and have come across a couple of things:
1. SSH support for `git-lfs` was added in 2021 for the v3.0.0 release: https://github.com/git-lfs/git-lfs/pull/4446
2. I've run a git trace and have discovered that, when I run `git push`, transferring the LFS data over pure SSH fails (probably because GitHub doesn't implement this protocol for LFS on the server-side), however `git-lfs` then falls back to authenticating over SSH using `git-lfs-authenticate` and from what I can gather this generates the `RemoteAuth gitauth-v1-XXXXXXXXXXXXX` token (I have redacted my actual token with Xs) and uses this to authenticate the payload that gets sent over HTTPS. So it seems to be a hybrid approach of authenticating over SSH to generate an auth token, and then sending the data over HTTPS using that auth token
```
23:58:03.028128 trace git-lfs: attempting pure SSH protocol connection
23:58:03.028163 trace git-lfs: spawning pure SSH connection
23:58:03.028459 trace git-lfs: run_command: ssh -oControlMaster=yes -oControlPath=/tmp/sock-646106498/lfs.sock git@github.com git-lfs-transfer RyanMan56/Tunnel-Crew.git upload
23:58:03.028684 trace git-lfs: exec: ssh '-oControlMaster=yes' '-oControlPath=/tmp/sock-646106498/lfs.sock' 'git@github.com' 'git-lfs-transfer RyanMan56/Tunnel-Crew.git upload'
23:58:03.914050 trace git-lfs: pure SSH connection successful
23:58:03.914143 trace git-lfs: pure SSH protocol connection failed: Unable to negotiate version with remote side (unable to read capabilities): EOF
23:58:03.914608 trace git-lfs: run_command: ssh git@github.com git-lfs-authenticate RyanMan56/Tunnel-Crew.git upload
23:58:03.914917 trace git-lfs: exec: ssh 'git@github.com' 'git-lfs-authenticate RyanMan56/Tunnel-Crew.git upload'
23:58:04.869444 trace git-lfs: exec: /usr/bin/security 'list-keychains'
23:58:04.885635 trace git-lfs: exec: /usr/bin/security 'find-certificate' '-a' '-p' '-c' 'lfs.github.com' '/Library/Keychains/System.keychain'
23:58:04.900620 trace git-lfs: HTTP: POST https://lfs.github.com/RyanMan56/Tunnel-Crew/locks/verify
> POST /RyanMan56/Tunnel-Crew/locks/verify HTTP/1.1
> Host: lfs.github.com
> Accept: application/vnd.git-lfs+json
> Authorization: RemoteAuth gitauth-v1-XXXXXXXXXXXXX
> Content-Length: 34
> Content-Type: application/vnd.git-lfs+json; charset=utf-8
> User-Agent: git-lfs/3.5.1 (GitHub; darwin arm64; go 1.23.1)
```

Out of curiosity, what version of `git-lfs` are you using? This is my version:
```
❯ git-lfs -v                                                                                                                                                 00:35:00  75%
git-lfs/3.5.1 (GitHub; darwin arm64; go 1.23.1)
```

**Comment by mathiasplans** [comment]
I also suggest adding *.tmp to .gitignore. They appear when saving the project while the game is playing.

**Reply by skplunkerin** [reply]
Is that for Windows? I just tested this on Mac and don't see any `.tmp` files created anywhere 🤔 (I'm using Github's Godot `.gitignore` template: https://github.com/github/gitignore/blob/main/Godot.gitignore)

**Reply by mathiasplans** [reply]
Yes, on WIndows. Seems that it was added to .gitignore three weeks ago, so it is solved :)

**Comment by louisnguyen12** [comment]
If you already configured SSH token on github, and the error still has "Is your credentials correct? Try using PAT token ...".  <br/>
**Try these steps**:
- Go to your github profile /Settings/Developer settings/Personal access tokens , then click on Generating new token / token (classic). 
- Set your token name
- Expiration (normally 30 days, you can set it longer if you want to)
- Select "Only select repositories" then select your repository.
- Change the Permission, I'm not sure which one to leave it Only Read or Read-and-Write (because of security reasons), here are few permissions I already set:
   - Repository permissions:
     * Actions: Read-and-Write
     * Adminstration: Read-and-Write
     * Codespaces: Read-and-Write
     * Contents: Read-and-Write
     * Issues: Read-and-Write
     * Merge queues: Read-and-Write
     * Pull requests: Read-and-Write
   - Account permissions:
     * Git SSH keys: Read-Only
     * SSH signing keys: Read-Only
The rest I don't know what they do so I just leave them be.
- Generate Token then Copy the token then paste it in your Project/Version Control/Version Control Settings/Password

Hope this helps.
